pkg/platform/platform_versioner.go

```diff
@@ -0,0 +1,140 @@
+package platform
+
+import (
+	"encoding/json"
+	"errors"
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var (
+	log                     = logf.Log.WithName("utils")
+	ErrInfoFetch            = errors.New("error fetching PlatformInfo")
+	ClusterVersionApiPath   = "apis/config.openshift.io/v1/clusterversions/version"
+	ErrDefaultingArgs       = errors.New("error while defaulting non-provided args for PlatformInfo fetch")
+	ErrRestConfigFetch      = errors.New("error fetching REST config to replace passed nil")
+	ErrDiscoveryClientFetch = errors.New("error getting new DiscoveryClient")
+	ErrK8SVersionFetch      = errors.New("error fetching K8S version from ServerVersion")
+	ErrServerGroupsFetch    = errors.New("error fetching ServerGroups from discovery client")
+	ErrOpenAPISchemaFetch   = errors.New("error fetching OpenAPISchema")
+	ErrClusterVersionFetch  = errors.New("error fetching OCP4 version via API")
+	ErrVersionMarshalling   = errors.New("error unmarshalling version API result to ClusterVersionInfo")
+)
+
+type PlatformVersioner interface {
+	GetPlatformInfo(discoverer Discoverer, cfg *rest.Config) (PlatformInfo, error)
+}
+
+type Discoverer interface {
+	ServerVersion() (*version.Info, error)
+	ServerGroups() (*v1.APIGroupList, error)
+	OpenAPISchema() (*openapi_v2.Document, error)
+	RESTClient() rest.Interface
+}
+
+type K8SBasedPlatformVersioner struct{}
+
+// deal with cfg coming from legacy method signature and allow injection for client testing
+func (K8SBasedPlatformVersioner) DefaultArgs(client Discoverer, cfg *rest.Config) (Discoverer, *rest.Config, error) {
+	if cfg == nil {
+		var err error
+		cfg, err = config.GetConfig()
+		if err != nil {
+			log.Error(err, ErrRestConfigFetch.Error())
+			return nil, nil, ErrRestConfigFetch
+		}
+	}
+	if client == nil {
+		var err error
+		client, err = discovery.NewDiscoveryClientForConfig(cfg)
+		if err != nil {
+			log.Error(err, ErrDiscoveryClientFetch.Error())
+			return nil, nil, ErrDiscoveryClientFetch
+		}
+	}
+	return client, cfg, nil
+}
+
+func (pv K8SBasedPlatformVersioner) GetPlatformInfo(client Discoverer, cfg *rest.Config) (PlatformInfo, error) {
+	log.Info("detecting platform version...")
+	info := PlatformInfo{Name: Kubernetes}
+
+	var err error
+	client, cfg, err = pv.DefaultArgs(client, cfg)
+	if err != nil {
+		log.Error(err, ErrDefaultingArgs.Error())
+		return info, ErrDefaultingArgs
```

Comment:
IMO, this qualifies as swallowing the exception. The caller should be able to programmatically inspect the underlying error and discern the issue. They would not be able to do so here, and they cannot (nor should) consume the log. Conversely, if the actual error is returned, it should be logged at the debug level since the caller makes the determination whether it's an error scenario for the their application, or something that might happen and they can recover from.

Reply:
Fair enough. I debated bubbling errors when I added this pattern - I think there could be value in appending something indicative of the calling code issue so that you get more of a stack trace effect for troubleshooting, but not at the cost of swallowing the underlying error. Would a better practice be to drop the logging altogether and use something like this instead?

`return nil, fmt.Errorf("could not fetch default REST config: %v", err)`

Reply:
I've altered the code (and testing) to use the above pattern I mentioned, please take a look. Feels cleaner to me and seems to adhere to patterns use in k8s code.

Reply:
I don't know, the original error is now being returned somewhere in the content, but is it usable? Maybe I'm thinking too much Java here, and it would help to think of a practical situation. I call your function, anticipating it might fail because of some type of security error. I want to check the error and see if that's what it was. Can I?

Reply:
per our chat convo, errors that occur are directly bubbled up to calling code now. Tests also adjusted to reflect change.